[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

CarlosAlbertoEnciso opened the pull request:
https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

llvm-debuginfo-analyzer already uses the tombstone approach
to identify dead code. Currently, the tombstone value is
evaluated as std::numeric_limits<uint64_t>::max(). Which is
wrong as it does not take into account the 'Address Byte Size'
from the Compile Unit header.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

llvm-debuginfo-analyzer already uses the tombstone approach
to identify dead code. Currently, the tombstone value is
evaluated as std::numeric_limits<uint64_t>::max(). Which is
wrong as it does not take into account the 'Address Byte Size'
from the Compile Unit header.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

Fix clang-format issue.

CarlosAlbertoEnciso commented on llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/wasm-crash-tombstone.test:
```diff
@@ -0,0 +1,44 @@
+; REQUIRES: x86-registered-target
```
> Do we usually split out the test and .s file in these tests? Most other tests in the LLVM project keep the content in the same file where possible.
> 
I just followed the same approach as for the existing .s files. I don't mind to combine the test and .s file.

> (& some comments describing what the assembly is from/what it describes might be handy too, either here (if everything can be merged), or in the assembly file (if it can't be merged))

I will update the comments.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

Address comments from reviewers:
- Combine test and .s file.
- Introduce the concept of a meta-tombstone value, to
  be able to check for an invalid tombstone value.

CarlosAlbertoEnciso commented:
Uploaded a new patch:
- Combine the test and .s file
- Introduced the 'meta-tombstone' concept

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

Minor comments change in test case.

CarlosAlbertoEnciso commented:
@dwblaikie, @Mrmaxmeier Thanks for your valuable feedback.
Is there any additional points about the current patch that you would like to address?

Thanks

dwblaikie reviewed (commented):
Test case could be a bit shorter/simpler? single CU with single function (with nothing, or a single nop/ret/something in it) with debug info describing it as dead?

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

Address reviewes comments:
- Reduce the test case.
- Use std::optional to illustrate the set/unset state.
- Rebase with main branch.

jalopezg-git commented on llvm/include/llvm/DebugInfo/LogicalView/Core/LVRange.h:
I don't think we need these changes to `LVRange`, as I have seen no constructor being called w/ a value different from the default (`std::nullopt`).

That would save space, as we don't need the `TombstoneAddress` additional member.  WDYT?

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

Address reviewes comments:
- Discard the changes done to 'LVRange.h'.

dwblaikie commented on llvm/test/tools/llvm-debuginfo-analyzer/WebAssembly/wasm-crash-tombstone.s:
```diff
@@ -0,0 +1,124 @@
+# llvm-debuginfo-analyzer crashes on dead code
+# https://github.com/llvm/llvm-project/issues/136772
+
+# For the attached reproducer:
+# llvm-dwarfdump out/lzma-lzmadec.wasm --all
+#
+# shows:
+#
+# 0x000002b3:   DW_TAG_subprogram
+#                 DW_AT_low_pc	(dead code)
+#                 DW_AT_high_pc	(0x00000362)
+#                 DW_AT_frame_base	(DW_OP_WASM_location 0x0 0x6, DW_OP_stack_value)
+
+# llvm-debuginfo-analyzer out/lzma-lzmadec.wasm --print=instructions
+#
+# crashes and shows a stack dump:
+#
+# PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/
+# and include the crash backtrace.
+# Stack dump:
+# 0. Program arguments: llvm-debuginfo-analyzer out/lzma-lzmadec.wasm --print=instructions
+
```
This commentary, and the name of the test file are probably not suitable/may be confusing - they describe the state of the world before this patch, and in somewhat vague terms "this complex test causes this thing to crash" rather than the specific issue it's been debugged down to.





Something more like "wasm-32bit-tombstone" might be a more suitable title, and a sentence describing "test that DWARF tombstones are correctly detected/respected in wasm 32 bit object files"? (& then proceed to describe the test case as done below - creating a single function, tombstoning it in the assembly, etc)

CarlosAlbertoEnciso replied:
Renamed the test case and reworked the comments to add clarity.

CarlosAlbertoEnciso commented on llvm/include/llvm/DebugInfo/LogicalView/Readers/LVCodeViewReader.h:
```diff
@@ -193,11 +193,17 @@ class LVCodeViewReader final : public LVBinaryReader {
       : LVBinaryReader(Filename, FileFormatName, W, LVBinaryType::COFF),
-        Input(&Obj), ExePath(ExePath), LogicalVisitor(this, W, Input) {}
+        Input(&Obj), ExePath(ExePath), LogicalVisitor(this, W, Input) {
+    // CodeView does not have the concept of 'tombstone' address.
```
That is a very good point. Moving the `tombstone` to the `LVDWARFReader` makes the approach very clean with no extra dependencies.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Fix crash with WebAssembly dead code

https://github.com/llvm/llvm-project/issues/136772
Incorrect handling of 'tombstone' value for WebAssembly.

Address reviewes comments:
- Make 'tombstone' a member of 'LVDWARFReader', to eliminate the
  extra dependency on the CodeView reader.
- Rename the test case and rework some comments to make it easy
  to understand.

CarlosAlbertoEnciso commented:
@dwblaikie Thanks for your input. Updated patch.

CarlosAlbertoEnciso commented:
The test case includes an incorrect `REQUIRES`.
It should have `# REQUIRES: webassembly-registered-target`

CarlosAlbertoEnciso commented:
Open https://github.com/llvm/llvm-project/pull/145848 that fixes the invalid `REQUIRES`.
